File: backend/sdk/handlers/narrative_tools.py
"""
Narrative output tools for TRPG gameplay.

Contains tools for presenting narrative content to the player:
- narration: Create visible narrative messages
- suggest_options: Provide suggested actions as clickable buttons

These tools handle the "presentation layer" - displaying content to the player
rather than modifying game state.
"""

import logging
from typing import Any

import crud
import schemas
from claude_agent_sdk import tool
from domain.value_objects.enums import MessageRole
from infrastructure.logging.perf_logger import track_perf
from services.player_service import PlayerService
from services.room_mapping_service import RoomMappingService

from sdk.handlers.context import ToolContext
from sdk.loaders import get_tool_description, is_tool_enabled
from sdk.tools.gameplay import (
    NarrationInput,
    SuggestOptionsInput,
)

logger = logging.getLogger("GameplayTools.Narrative")


def create_narrative_tools(ctx: ToolContext) -> list:
    """
    Create narrative output tools.

    Args:
        ctx: Unified tool context containing agent info and dependencies

    Returns:
        List of narrative tool functions
    """
    tools = []

    # Get required dependencies from context
    db = ctx.require_db()
    world_name = ctx.require_world_name()

    # ==========================================================================
    # narration tool - Create visible narrative message (replaces Narrator)
    # ==========================================================================
    if is_tool_enabled("narration", default=True):
        narration_description = get_tool_description(
            "narration",
            agent_name="Action Manager",
            group_name=ctx.group_name,
            default="REQUIRED: Create a visible narrative message describing the outcome. "
            "This is the text the player will see. Make it vivid and engaging.",
        )

        @tool(
            "narration",
            narration_description,
            NarrationInput.model_json_schema(),
        )
        @track_perf(
            "tool_narration",
            room_id=lambda: ctx.room_id,
            agent_name=lambda: ctx.agent_name,
            include_result=True,
        )
        async def narration_tool(args: dict[str, Any]):
            """
            Create a visible narrative message in the conversation.

            This replaces the separate Narrator agent. Action Manager now
            handles both interpretation and narration in a single turn.
            """
            # Validate input with Pydantic
            validated = NarrationInput(**args)
            narrative = validated.narrative

            logger.info(f"📖 narration called | length={len(narrative)}")

            try:
                # Get Action Manager's agent_id and room_id from context
                agent_id = ctx.require_agent_id()
                room_id = ctx.require_room_id()

                # Load game_time_snapshot from player state for message timestamp
                game_time_snapshot = None
                player_state = PlayerService.load_player_state(world_name)
                if player_state and player_state.game_time:
                    game_time_snapshot = player_state.game_time

                # Build thinking field with NPC reactions (for collapsible display)
                thinking_content = None
                if ctx.npc_reactions:
                    parts = ["[NPC_REACTIONS]"]
                    for reaction in ctx.npc_reactions:
                        parts.append(f"=== {reaction['agent_name']} ===")
                        parts.append(reaction["content"])
                        parts.append("")
                    parts.append("[/NPC_REACTIONS]")
                    thinking_content = "\n".join(parts)
                    logger.info(f"📝 Including {len(ctx.npc_reactions)} NPC reactions in thinking field")

                # Create message via crud.create_message()
                message = schemas.MessageCreate(
                    content=narrative,
                    role=MessageRole.ASSISTANT,
                    agent_id=agent_id,
                    game_time_snapshot=game_time_snapshot,
                    thinking=thinking_content,  # Store NPC reactions in thinking field
                )
                await crud.create_message(db, room_id, message, update_room_activity=True)

                # Signal that narration has been produced (allows input unblocking)
                # Lazy import to avoid circular dependency
                from orchestration.trpg_orchestrator import get_trpg_orchestrator

                get_trpg_orchestrator().set_narration_produced(room_id)

                logger.info(f"✅ Narrative message created | room={room_id} | agent={agent_id}")

                return {
                    "content": [
                        {
                            "type": "text",
                            "text": "✓ Narrative message created and displayed to player.",
                        }
                    ]
                }

            except Exception as e:
                logger.error(f"narration error: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error creating narrative: {e}"}],
                    "is_error": True,
                }

        tools.append(narration_tool)

    # ==========================================================================
    # suggest_options tool - Provide suggested actions for player
    # ==========================================================================
    if is_tool_enabled("suggest_options", default=True):
        suggest_options_description = get_tool_description(
            "suggest_options",
            agent_name="Action Manager",
            group_name=ctx.group_name,
            default="REQUIRED: Provide two suggested actions for the player. "
            "These appear as clickable buttons in the UI.",
        )

        @tool(
            "suggest_options",
            suggest_options_description,
            SuggestOptionsInput.model_json_schema(),
        )
        @track_perf(
            "tool_suggest_options",
            room_id=lambda: ctx.room_id,
            agent_name=lambda: ctx.agent_name,
            include_result=True,
        )
        async def suggest_options_tool(args: dict[str, Any]):
            """
            Provide two suggested actions for the player.

            These suggestions appear as clickable buttons in the UI.
            Replaces the Narrator's suggest_actions tool.
            """
            # Validate input with Pydantic
            validated = SuggestOptionsInput(**args)
            action_1 = validated.action_1
            action_2 = validated.action_2

            logger.info(f"💡 suggest_options invoked: [{action_1}] / [{action_2}]")

            if not action_1 or not action_2:
                return {
                    "content": [{"type": "text", "text": "Please provide both action suggestions."}],
                    "is_error": True,
                }

            # Persist suggestions to _state.json for frontend retrieval
            try:
                RoomMappingService.save_suggestions(world_name, [action_1, action_2])
                logger.info(f"💾 Suggestions persisted to _state.json for world: {world_name}")
            except Exception as e:
                logger.error(f"Failed to persist suggestions: {e}")

            # Format response for the system
            response_text = f"""**Suggested Actions:**
1. {action_1}
2. {action_2}"""

            return {
                "content": [{"type": "text", "text": response_text}],
            }

        tools.append(suggest_options_tool)

    return tools

File: backend/src/sdk/handlers/narrative-tools.ts
import { getAgent } from '@/crud/agents'
import { createMessage, type MessageWithAgent } from '@/crud/messages'
import type { PlayerService } from '@/services/player-service'
import type { RoomMappingService } from '@/services/room-mapping'
import { formatTemplate } from '@/sdk/tools/definitions'
import { awaitReactionsTool, narrationTool, suggestOptionsTool } from '@/sdk/tools/gameplay'
import { tool, requireAgentId, requireRoomId, requireWorldName, toolSuccess, type SdkTool, type ToolContext } from './context'

// The tools that produce what the player actually sees.

export interface NarrativeDeps {
  players: PlayerService
  rooms: RoomMappingService
  /** Called when narration lands, so the turn loop knows the player can act again. */
  onNarrationProduced?: (roomId: number) => void
  /**
   * Push the saved line to the room's SSE clients.
   *
   * `turn.ts` fires its own `onMessageSaved`, but never for the Action Manager:
   * that agent is hidden, so the turn loop persists nothing and this tool is
   * the only thing that writes what the player reads. Without this the finished
   * narration waits for whenever the next poll lands.
   */
  onNarrationSaved?: (roomId: number, message: MessageWithAgent) => void
}

/**
 * Wrap NPC reactions for the narration message's `thinking` column — the only
 * durable record of them, since reactions are never persisted as messages. The
 * frontend keys on these markers to render the collapsible.
 */
export function serializeNpcReactions(
  reactions: ReadonlyArray<{ agentName: string; content: string }>,
): string | null {
  if (reactions.length === 0) return null
  const body = reactions.map((r) => `=== ${r.agentName} ===\n${r.content}`).join('\n\n')
  return `[NPC_REACTIONS]\n${body}\n[/NPC_REACTIONS]`
}

export function createNarrativeTools(
  ctx: ToolContext,
  deps: NarrativeDeps,
): SdkTool[] {
  // Resolved at build time, not call time: a tool whose context is incomplete
  // must not be offered to the model at all.
  const roomId = requireRoomId(ctx)
  const agentId = requireAgentId(ctx)
  const worldName = requireWorldName(ctx)

  const narration = tool(
    narrationTool.name,
    narrationTool.description,
    narrationTool.inputSchema,
    async (args) => {
      const narrative = args.narrative
      const playerState = deps.players.loadPlayerState(worldName)

      const db = ctx.getDb()
      const saved = createMessage(db, roomId, {
        content: narrative,
        role: 'assistant',
        agentId,
        thinking: serializeNpcReactions(ctx.npcReactions ?? []),
        gameTimeSnapshot: playerState?.gameTime ?? null,
      })
      deps.onNarrationSaved?.(roomId, { ...saved, agent: getAgent(db, agentId) })

      // Load-bearing for turn flow: the signal that input can reopen.
      deps.onNarrationProduced?.(roomId)

      return toolSuccess(narrationTool.response ?? 'Narrative message created and displayed to player.')
    },
  )

  // Always offered, in flight or not: tool sets must not vary per turn — the
  // allow-list is baked into the session at `query()` time — so an empty
  // location gets the tool and a sentence saying there was nobody to hear it.
  const awaitReactions = tool(
    awaitReactionsTool.name,
    awaitReactionsTool.description,
    awaitReactionsTool.inputSchema,
    async () => {
      const reactions = (await ctx.awaitNpcReactions?.()) ?? ctx.npcReactions ?? []
      return toolSuccess(
        formatTemplate(awaitReactionsTool.response ?? '{reactions}', {
          reactions: formatReactions(reactions),
        }),
      )
    },
  )

  const suggestOptions = tool(
    suggestOptionsTool.name,
    suggestOptionsTool.description,
    suggestOptionsTool.inputSchema,
    async (args) => {
      const { action_1, action_2 } = args
      // Suggestions live in the world's `_state.json`, not the database.
      deps.rooms.saveSuggestions(worldName, [action_1, action_2])
      return toolSuccess(
        formatTemplate(suggestOptionsTool.response ?? '', { action_1, action_2 }),
      )
    },
  )

  return [narration, awaitReactions, suggestOptions]
}

/**
 * What the Action Manager reads back. Each reaction is reproduced whole — this
 * is the only copy that exists, since a reaction is never persisted as a message
 * — and the closing line is there because a model handed a block of NPC prose
 * with no instruction summarises it instead of quoting it.
 */
function formatReactions(reactions: ReadonlyArray<{ agentName: string; content: string }>): string {
  if (reactions.length === 0) {
    return 'No one else is at this location — nobody reacted. Narrate the scene as the player alone in it.'
  }

  const blocks = reactions.map((r) => `### ${r.agentName}\n${r.content}`)
  return [
    `${String(reactions.length)} character(s) reacted to the player's action:`,
    '',
    ...blocks,
    '',
    'Now call `narration` for this: name each character, quote what they said as spoken ' +
      'dialogue and stage what they did. These lines reach the player only through you.',
  ].join('\n')
}
